crates/amaru-kernel/src/lib.rs

```diff
@@ -783,6 +788,63 @@ pub fn to_ex_units(witness_set: WitnessSet) -> ExUnits {
     }
 }
 
+/// Collect provided scripts and compute each ScriptHash in a witness set
+pub fn get_provided_scripts(witness_set: &MintedWitnessSet<'_>) -> BTreeSet<ScriptHash> {
+    let mut provided_scripts: BTreeSet<ScriptHash> = BTreeSet::new();
+    provided_scripts.extend(
```

Comment:
Might be worth considering [chain](https://doc.rust-lang.org/std/iter/struct.Chain.html) as a way to reduce the overhead of multiple `extends` now probably not worth from a resource constraints here. More as a nice exercise towards a more functional approach :)